### ISSUE TITLE
Messages must be consumed before another Message is published

### ISSUE THREAD

**Issue by twahage** [fridujo/rabbitmq-mock#96]
For a spring boot test case i do send multiple messages to an exchange routing to an single queue in a test setup.
While verifying the messages in the queue only the first message can be consumed. Any further rabbitTemplate.receive() calls for the remaining messages deliver null (with a sufficient timeout). The tests do work though if i immediately  consume messages after sending before publishing the next one. Regardless of this behavior being intended... it might be useful to be able to queue multiple messages at once

**Comment by ledoyen** [comment]
Hello,

Thanks for reporting this.
Could you provide more details, ideally a JUnit test to demonstrate the behavior ?

I tried to reproduce what you describes in #97 but it seems to work.

**Comment by twahage** [comment]
Hello,

thanks for the quick reply! 
Sure, below is a unit test that does reproduce the error. Details about the Environment are 

- java 8
- spring-boot-starter-parent: 2.1.7.RELEASE
- amqp-client: 5.4.3
- and using junit 5

I did try it on another project though with following specs, which actually works: 

- java 11
- spring-boot-starter-parent: 2.1.9.RELEASE
- amqp-client: 5.4.3
- and also using junit 5 

```java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.fridujo.rabbitmq.mock.MockConnectionFactory;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.annotation.EnableRabbit;
import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.amqp.RabbitAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest(classes = {RabbitAutoConfiguration.class, MockedRabbitMQTest.MockedRabbitConfiguration.class})
public class MockedRabbitMQTest {

    private static final long RABBIT_RECEIVE_OP_TIMEOUT_IN_MILLIS = 1000;

    private static final String MOCKED_QUEUE = "mocked_queue";

    private static final String ROUTING_KEY = "test.message";

    private static final String MOCKED_EXCHANGE = "mocked_exchange";

    @Autowired
    private AmqpAdmin amqpAdmin;

    @Autowired
    private RabbitTemplate rabbitTemplate;

    private static final ObjectMapper objectMapper = new ObjectMapper();

    @BeforeEach
    public void setup() {
        Queue queue = QueueBuilder.durable(MOCKED_QUEUE).build();
        Exchange exchange = ExchangeBuilder.topicExchange(MOCKED_EXCHANGE).build();
        amqpAdmin.declareExchange(exchange);
        amqpAdmin.declareQueue(queue);
        Binding binding = BindingBuilder
                .bind(queue)
                .to(exchange)
                .with(ROUTING_KEY.trim())
                .noargs();
        amqpAdmin.declareBinding(binding);
    }

    @AfterEach
    public void tearDown() {
        Queue queue = QueueBuilder.durable(MOCKED_QUEUE).build();
        Exchange exchange = ExchangeBuilder.topicExchange(MOCKED_EXCHANGE).build();
        Binding binding = BindingBuilder
                .bind(queue)
                .to(exchange)
                .with(ROUTING_KEY.trim())
                .noargs();
        amqpAdmin.removeBinding(binding);
        amqpAdmin.deleteQueue(MOCKED_QUEUE);
    }

    @Test
    public void rabbitMockTest() throws IOException {
        JsonNode event = createNewTrackedTimeEvent();

        Message messageToSend = MessageBuilder
                .withBody(event.toString().getBytes())
                .setContentType(MessageProperties.CONTENT_TYPE_JSON)
                .setDeliveryMode(MessageDeliveryMode.PERSISTENT)
                .build();

        rabbitTemplate.convertAndSend(MOCKED_EXCHANGE, ROUTING_KEY, messageToSend);
        rabbitTemplate.convertAndSend(MOCKED_EXCHANGE, ROUTING_KEY, messageToSend);
        rabbitTemplate.convertAndSend(MOCKED_EXCHANGE, ROUTING_KEY, messageToSend);

        Message messageA = rabbitTemplate.receive(MOCKED_QUEUE, RABBIT_RECEIVE_OP_TIMEOUT_IN_MILLIS);
        Message messageB = rabbitTemplate.receive(MOCKED_QUEUE, RABBIT_RECEIVE_OP_TIMEOUT_IN_MILLIS);
        Message messageC = rabbitTemplate.receive(MOCKED_QUEUE, RABBIT_RECEIVE_OP_TIMEOUT_IN_MILLIS);

        assertEquals(event, objectMapper.readTree(messageA.getBody()));
        assertEquals(event, objectMapper.readTree(messageB.getBody()));
        assertEquals(event, objectMapper.readTree(messageC.getBody()));
    }

    private JsonNode createNewEvent() throws IOException {
        return objectMapper.readTree("{\n" +
                "  \"meta\": {\n" +
                "    \"lorem\": \"ipsum\"\n" +
                "  }\n" +
                "}");
    }

    @EnableRabbit
    @TestConfiguration
    static class MockedRabbitConfiguration {
        @Bean
        ConnectionFactory connectionFactory() {
            return new CachingConnectionFactory(new MockConnectionFactory());
        }
    }

}
```

**Comment by ledoyen** [comment]
Sorry for the late response, I will add these cases in IT as soon as I can.

Thanks for providing test-cases !

**Comment by ledoyen** [comment]
Hi,

It's a bug !
When using `RabbitTemplate#receive(String, long)` **Spring-Rabbit** is in fact creating a consumer.
This Consumer will get delivered with all messages available in that queue.
**Spring-Rabbit** will then ack the first one, as it is only interested in one message, then it cancels the consumer.

However when cancelled, unacked messages stay unacked, whereas they should become available anew.

**Comment by twahage** [comment]
Thanks for the information and your effort!

I will eagerly await that fix :)
In the meanwhile i would also try to find some fix or workaround and keep you up to date over this communication channel.

**Comment by ledoyen** [comment]
@twahage this should be fixed in release **1.1.0**
